security/fma-contract-version-upgrade.md

```diff
@@ -0,0 +1,122 @@
+# [Project Name]: Failure Modes and Recovery Path Analysis
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Introduction](#introduction)
+- [Failure Modes and Recovery Paths](#failure-modes-and-recovery-paths)
+  - [Missing side effect on `.selector` access bug (Bug fixed after 0.8.15)](#missing-side-effect-on-selector-access-bug-bug-fixed-after-0815)
+  - [Storage write removal before conditional termination (Bug fixed after 0.8.15)](#storage-write-removal-before-conditional-termination-bug-fixed-after-0815)
+  - [Change in Default EVM version between different Solidity versions](#change-in-default-evm-version-between-different-solidity-versions)
+- [Audit Requirements](#audit-requirements)
+- [Action Items](#action-items)
+- [Appendix](#appendix)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+|                     |               |
+| ------------------- | ------------- |
+| Author              | Michael Amadi |
+| Created at          | 2024-08-20    |
+| Needs Approval From | Matt Solomon  |
+| Other Reviewers     | Mark Tyneway  |
+| Status              | Draft         |
+
+> [!NOTE]
+> 📢 Remember:
+>
+> - The single approver in the “Need Approval From” must be from the Security team.
+> - Maintain the “Status” property accordingly.
+
+> [!TIP]
+> Guidelines for writing a good analysis, and what the reviewer will look for:
+>
+> - Show your work: Include steps and tools for each conclusion.
+> - Completeness of risks considered.
+> - Include both implementation and operational failure modes
+> - Provide references to support the reviewer.
+> - The size of the document will likely be proportional to the project's complexity.
+> - The ultimate goal of this document is to identify action items to improve the security of the project. The FMA review process can be accelerated by proactively identifying action items during the writing process.
+
+## Introduction
+
+This document covers solidity version upgrade of all OP Stack smart contracts from 0.8.15 code to 0.8.25.
+
+Below are references for this project:
+
+- [Issue 11527: Smart Contracts: Update all 0.8.15 code to 0.8.25](https://github.com/ethereum-optimism/optimism/issues/11527).
+
+## Failure Modes and Recovery Paths
```

Comment:
Can you add some intro text to this section explaining why the below solidity bug fixes are present? In other words, there have likely been a lot of bug fixes since 0.8.15, and we should definitely not enumerate all of them here, so the text would explain why this subset is the one listed.

I'd also suggest a slight tweak to the format:

```markdown
## Failure Modes and Recovery Paths

### Solidity Bugs Fixed

This section lists solidity bugs that were fixed as failure modes and explains why these fixes might be a risk for us. This set of bug fixes was chosen because....

#### Missing side effect on `.selector` access bug (Bug fixed after 0.8.15)

etc...

### Other Failure Modes

#### New bugs introduced in later solidity versions

#### Stack too deep

etc....
```

Reply:
The above comment also hints at another suggestion: we should have a failure mode for exacerbating our [stack too deep](https://github.com/ethereum-optimism/optimism/issues/11093) problem. Currently our codebase only compile with the legacy pipeline with optimizer on. We cannot compile with optimizer off, nor with via-ir, as both result in stack too deep. We need to make sure whatever version we choose does not also give us stack too deep for our current build settings